### PR TITLE
bluestore: Elastic Shared Blobs 4 - tests

### PR DESCRIPTION
**Pull request by aclamk**
This is part 4/4 of ESB work.
It encompasses all tests that had been modified / added for this feature.
#51439 
#51440 
#51441 
#51442

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by aclamk** [comment]
jenkins test make check

**Comment by aclamk** [comment]
jenkins test make check

**Comment by aclamk** [comment]
jenkins test make check

**Comment by aclamk** [comment]
jenkins test make check

**Comment by aclamk** [comment]
jenkins test make check

**Comment by ifed01** [comment]
jenkins test make check

File: src/test/objectstore/test_bluestore_types.cc [hunk 1]
```diff
@@ -1284,6 +1284,380 @@ TEST(ExtentMap, compress_extent_map)
...
+  }
+
+  void write(t_onode& onode, uint32_t off, uint32_t len) {
+    ceph_assert((off % au_size) == 0);
+    ceph_assert((len % au_size) == 0);
+    punch_hole(onode, off, len);
+
+    uint32_t i = off / au_size;
+    uint32_t end = (off + len) / au_size;
+    fillup(onode, end);
```
Comment (ifed01) [inline src/test/objectstore/test_bluestore_types.cc:1401]: redundant? punch_hole does the same...

File: src/test/objectstore/test_bluestore_types.cc [hunk 2]
```diff
@@ -1284,6 +1284,380 @@ TEST(ExtentMap, compress_extent_map)
...
+    // below simulation of write performed by BlueStore::do_write()
+    auto helper_blob_write = [&](
+      uint32_t log_off,   // logical offset of blob to put to onode
+      uint32_t empty_aus, // amount of unreferenced aus in the beginning
+      uint32_t first_au,  // first au that will be referenced
+      uint32_t num_aus     // number of aus, first, first+1.. first+num_au-1
+    ) {
+      uint32_t blob_length = (empty_aus + num_aus) * au_size;
+      BlueStore::BlobRef b(new BlueStore::Blob);
+      b->shared_blob = new BlueStore::SharedBlob(coll.get());
```
Comment (ifed01) [inline src/test/objectstore/test_bluestore_types.cc:1421]: use set_shared_blob?

**Review by ifed01 — APPROVED** [review]
'tests' part LGTM.  Just a couple nits...

**Comment by ifed01** [comment]
jenkins test make check

**Comment by rzarzynski** [comment]
@ljflores, @aclamk, @yuriw: IIUC it's enough to put under QA only the 4th part (this PR) of the ESB series as it has all the commits from:

* https://github.com/ceph/ceph/pull/51439
* https://github.com/ceph/ceph/pull/51440
* https://github.com/ceph/ceph/pull/51441

Let's go through the official QA, merge and bake it in `main`!

**Comment by ljflores** [comment]
Hey @aclamk, sorry for the delay. This is rados-approved! https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcWt1KTViI1830-wip-yuri-testing-2023-08-25-0809

**Comment by ljflores** [comment]
@rzarzynski FYI ^

I'll leave it up to you guys to merge if it's ready.

**Comment by rzarzynski** [comment]
Merged as a part of https://github.com/ceph/ceph/pull/53178.

**Pull request merged**